### DISCUSSION TITLE
CWEB test suite?

### DISCUSSION THREAD

**Discussion by shreevatsa** [shreevatsa/webWEB#4] [General]
(Created this as a new "topic" for ease of finding later… Sorry for the noise; it got created as an "issue" first and then had to be changed into a "discussion".) 

_Originally posted by @zmajeed in https://github.com/shreevatsa/webWEB/discussions/1#discussioncomment-873270_ 

> Hey - I'm testing my cweb grammar and parser with graphbase files - I also have mmixware files and Knuth's programs from his website - still I'd appreciate links to any other cweb files especially ones with complex macros that contain cweb control texts - if you know of a cweb test suite that would be the best. Thanks!

**Comment by shreevatsa** [comment]
I think the ones mentioned in the question are already a great start, worth listing out again for future reference:

- [Stanford GraphBase](https://www-cs-faculty.stanford.edu/~knuth/sgb.html), specifically [`sgb.tar.gz`](http://ftp.cs.stanford.edu/pub/sgb/sgb.tar.gz)
- [MMIXWare](https://www-cs-faculty.stanford.edu/~knuth/mmixware.html), specifically [`mmix-20131017.tgz`](https://www-cs-faculty.stanford.edu/~knuth/programs/mmix-20131017.tgz)?
- [Programs from Knuth's website](https://www-cs-faculty.stanford.edu/~knuth/programs.html)

And a suggestion by @texdraft:
- [web2w](https://w3-o.cs.hm.edu/users/ruckert/public_html/web2w/index.html) ([CTAN](https://ctan.org/tex-archive/web/web2w))

And of course:
- The [programs](https://mirrors.ctan.org/info/knuth-pdf/index.html#cweb-programs-by-knuth-et-al.) that are part of CWEB itself: 
  > cweb/ [ COMMON CTANGLE CWEAVE CTWILL REFSORT TWINX]
  > ctie/ CTIE
  > tie/ TIE

  (See the repo https://github.com/ascherer/cweb for more?)

**Reply by texdraft** [reply]
Martin Ruckert also has an MP3-file-processing library whose `CWEB` documentation was published as a book. The website is [here](https://w3-o.cs.hm.edu/users/ruckert/public_html/understanding_mp3/), but the source  code download link doesn't seem to work.

**Reply by ascherer** [reply]
Note that [tie](https://www.tug.org/svn/texlive/trunk/Build/source/texk/web2c/tiedir) and [ctie](https://www.tug.org/svn/texlive/trunk/Build/source/texk/web2c/ctiedir) come from [TeX Live](https://github.com/TeX-Live/texlive-source) and are used for both `CWEB` and `WEB` sources. There you also find the `CWEB` codes for [MPlib](https://github.com/TeX-Live/texlive-source/tree/trunk/texk/web2c/mplibdir), but these are not in a state for (easy) publication.

**Reply by shreevatsa** [reply]
The MPlib looks like a good test case potentially. It is probably ok if that's not in a state for publication, as @zmajeed is not aiming for publication but simply to parse the cweb files into "regular" C code (replacing ctangle not cweave).

**Reply by zmajeed** [reply]
mplib has some very large files which is great - thanks! It has some unusual constructs like #define macros inside a named section with escaped newlines. A few nested named sections though not much deep nesting. I think I have enough variety now - I can also construct test cases based on these examples

It's true my initial interest was a custom and flexible ctangle - but now I'm going for a full grammar, parser and API - I hope to make any valid cweb file programmatically accessible - so people can develop tools like cweave and other cweb applications

**Reply by texdraft** [reply]
@zmajeed, I have a question: When you are referring to parsing CWEB, are you including parsing the C/C++ code?

**Reply by zmajeed** [reply]
not C/C++ source initially - I do need to parse comments because they are considered tex content - and some cweb format control codes are allowed in C/C++ source - as it happens Knuth provides a C/C++ grammar used by cweave - I'm not  sure my parser and API even need to fully tokenize C/C++ code - this may be better left to the application as long as the API provides all the info needed to extract the original source (along with the embedded format codes)

**Reply by texdraft** [reply]
I recently added some C++ syntax support to CWEAVE, and I can say that I would be very surprised/impressed if you managed to use CWEAVE's grammar for any actual parsing&hellip; Leaving everything that isn't CWEB-specific up to the application sounds like a good plan.

**Comment by dahosek** [comment]
I wonder whether it might make sense to integrate with LSP (https://microsoft.github.io/language-server-protocol/ <https://microsoft.github.io/language-server-protocol/>) for parsing the non-web parts of the code. Possibly, the web-parsing itself might be implemented with LSP (although I’ve not looked closely enough to know). LSP is primarily geared towards IDE development, but I imagine it could be repurposed for the sort of parsing that’s necessary for web tooling. Then, we’d get free IDE support for web and the ability to make webs for any arbitrary language all in one bundle.

-dh

**Reply by zmajeed** [reply]
yes - that's another reason to let an application handle program code - there could be existing parsers better suited to the application environment - also supposedly cweb can be used with other languages so it's better if my API remains language agnostic to the extent possible - I think that's what dahosek also points out

**Comment by zmajeed** [comment]
I've created a repo for a draft grammar, parser and API for cweb. There is a grammar available to discuss at https://github.com/gogotrango/cwebdraftapi/discussions/4 - any feedback on the grammar will be much appreciated.

I'll continue to participate in the discussions here and keep you posted on the API development
